### PR TITLE
Support for data-parallelism for equal algorithm

### PR DESCRIPTION
**Pull request by srinivasyadav18**
Contributes to fixing #2333

## Proposed Changes

  - Creates CPO for sequential_equal
  - Adapts datapar equal
  - Adds datapar equal unit tests

### CODE REVIEW

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/find.hpp [hunk 1]
```diff
@@ -381,7 +381,7 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
 
     template <typename ExPolicy, typename Iter1, typename Iter2, typename Token,
         typename Pred, typename Proj1, typename Proj2>
-    HPX_HOST_DEVICE HPX_FORCEINLINE Iter1 sequential_find_end(Iter1 it,
+    HPX_HOST_DEVICE HPX_FORCEINLINE void sequential_find_end(Iter1 it,
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/detail/find.hpp:384]: Why return `void` here? Shouldn't `find_end` provide the iterator to the found element?

File: libs/core/algorithms/include/hpx/parallel/algorithms/detail/find.hpp [hunk 2]
```diff
@@ -465,7 +465,7 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
 
     template <typename ExPolicy, typename FwdIter, typename FwdIter2,
         typename Token, typename Pred, typename Proj1, typename Proj2>
-    HPX_HOST_DEVICE HPX_FORCEINLINE FwdIter sequential_find_first_of(FwdIter it,
+    HPX_HOST_DEVICE HPX_FORCEINLINE void sequential_find_first_of(FwdIter it,
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/detail/find.hpp:468]: Same here, why the `void` return?

**Comment by hkaiser** [comment]
The HIP errors are unrelated. I created a ticket for those to be fixed separately: https://github.com/STEllAR-GROUP/hpx/issues/5764

File: libs/core/algorithms/include/hpx/parallel/algorithms/equal.hpp [hunk 3]
```diff
@@ -389,19 +367,12 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
                               std::size_t part_count) mutable -> bool {
-                    util::loop_n<std::decay_t<ExPolicy>>(it, part_count, tok,
-                        [&f, &tok](zip_iterator const& curr) mutable -> void {
-                            reference t = *curr;
-                            if (!HPX_INVOKE(f, hpx::get<0>(t), hpx::get<1>(t)))
-                            {
-                                tok.cancel();
-                            }
-                        });
+                    sequential_equal<std::decay_t<ExPolicy>>(
```
Comment (srinivasyadav18) [inline libs/core/algorithms/include/hpx/parallel/algorithms/equal.hpp:374]: Here, this lambda is now gets accessed by tag dispatch by `sequential_equal CPO`

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Pull request merged**